relations/src/r1cs/constraint_system.rs

```diff
@@ -91,6 +95,16 @@ pub enum SynthesisMode {
     },
 }
 
+/// Defines the parameter to optimize for a `ConstraintSystem`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum OptimizationGoal {
```

Comment:
Should we have a `None` here, which will do no optimization passes?

Reply:
That makes sense. Do you think `None`, or `Constraints`, should be the default?

Reply:
I think constraints should be the default

Reply:
What are the semantics of calling finalize with None? In the end we have to get rid of all SymbolicLC before converting to matrices

Reply:
I think `None` would still need to inline Symbolic LC (`self.inline_all_lcs()`, without which `to_matrices` would fail.

Dev, what do you think? Basically, at this moment, `None` would be the same as `Constraints`, and it is uncertain if in the future we will treat them differently.

Reply:
Right, but we wouldn't do any additional passes for inlining or outlining. Not sure if theres a real use case for that though. (Maybe it depends on if in-lining / out-lining is expensive relative to circuit generation in any context?)

Reply:
I think given that no specific use cases pop up for `None`, and we currently do not have a good idea of faster inlining (without optimization to reduce constraints), maybe we can omit `None` and add it later. Dev, what do you think?

Reply:
Sounds good to me

Reply:
Actually, I do think `None` is still useful, in that it could just make far fewer symbolic LC's. E.g. when multiplying by a constant, we don't need to make a new symbolic LC if we don't care about optimizations

Reply:
And we could have drop semantics to symbolic LC's to remove intermediate variables from the map as soon as they're dropped from memory. It is a significant engineering effort though, so maybe not worth having it now. (Its non breaking to add it later)

Reply:
That makes sense. Then let us keep `None` and, for now, let `None` also do `inline_all_lcs`. More changes may be coming in the future.